File: Leetcode/Medium/894-All_Possible_Full_Binary_Trees.cpp
#include <vector>
//Definition for a binary tree node.
struct TreeNode
{
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

vector<TreeNode *> allPossibleFBT(int n)
{
	if (n % 2 == 0)
		return {};
	vector<TreeNode *> cache[n + 1];
	cache[1] = {new TreeNode()};

	for (int place = 3; place <= n; place += 2) // starting place from 3 because for even numbers there are no trees and thus I am also skipping by two
		for (int i = 0; i < place; i++)
			for (auto left : cache[i])
				for (auto right : cache[place - i - 1])
					cache[place].push_back(new TreeNode(0, left, right));
	return cache[n];
}

File: Leetcode/Medium/894_All_Possible_Full_Binary_Trees.py
#Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def __init__(self):
        self.cache = {}
        self.cache[0] = []
        self.cache[1] = [TreeNode()]
    def allPossibleFBT(self, n: int) -> List[TreeNode]:# this could be done using tobulation
        if n%2==0: return []
        if n in self.cache: return self.cache[n]
        
        
        ans = []
        for i in range(1,n,2):
            left = self.allPossibleFBT(i)
            right = self.allPossibleFBT(n-i-1)
            for l in left:
                for r in right:
                    ans.append(TreeNode(0,l,r))
        self.cache[n] = ans
        return ans



# tobulation

class Solution:
    def allPossibleFBT(self, n: int) -> List[TreeNode]:
        if n%2==0: return []
        cache = [[] for _ in range(n+1)]
        cache[1].append(TreeNode())
        
        for place in range(3,n+1,2):
            for i in range(1,place,2):
                for left in cache[i]:
                    for right in cache[place-1-i]:
                        cache[place].append(TreeNode(0,left,right))
        return cache[n]

